File: Sources/ClipboardManager/Services/DatabaseService.swift
import Foundation
import SQLite3

/// SQLite 数据库服务：增删改查 + 自动清理
/// 与 Python 版使用相同的数据库 schema，数据可互通
final class DatabaseService {
    static let shared = DatabaseService()

    private var db: OpaquePointer?
    private let queue = DispatchQueue(label: "com.clipboardmanager.db", qos: .utility)

    // MARK: — 路径

    private var dataDir: String {
        let path = NSHomeDirectory() + "/Library/Application Support/ClipboardManager"
        try? FileManager.default.createDirectory(
            atPath: path, withIntermediateDirectories: true)
        return path
    }

    private var dbPath: String {
        dataDir + "/history.db"
    }

    var imagesDir: String {
        let path = dataDir + "/images"
        try? FileManager.default.createDirectory(
            atPath: path, withIntermediateDirectories: true)
        return path
    }

    // MARK: — 初始化

    private init() {
        openDatabase()
        createTables()
    }

    deinit {
        if let db = db {
            sqlite3_close(db)
        }
    }

    private func openDatabase() {
        guard sqlite3_open(dbPath, &db) == SQLITE_OK else {
            print("[DatabaseService] 无法打开数据库: \(dbPath)")
            return
        }
        // WAL 模式提升并发性能
        sqlite3_exec(db, "PRAGMA journal_mode=WAL", nil, nil, nil)
        print("[DatabaseService] 数据库已打开: \(dbPath)")
    }

    private func createTables() {
        let sql = """
            CREATE TABLE IF NOT EXISTS clipboard_items (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                content_type TEXT NOT NULL,
                text_content TEXT,
                image_path TEXT,
                text_hash TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                is_pinned INTEGER DEFAULT 0
            );
            CREATE INDEX IF NOT EXISTS idx_created_at
                ON clipboard_items(created_at DESC);
            CREATE INDEX IF NOT EXISTS idx_pinned
                ON clipboard_items(is_pinned);
        """
        execute(sql: sql)
    }

    // MARK: — 执行 SQL

    @discardableResult
    private func execute(sql: String) -> Bool {
        guard let db = db else { return false }
        return sqlite3_exec(db, sql, nil, nil, nil) == SQLITE_OK
    }

    // MARK: — 保存记录

    /// 保存一条剪贴板记录。返回新记录的 ID，如果重复则返回 nil。
    func saveItem(contentType: ClipboardItem.ContentType,
                  textContent: String? = nil,
                  imagePath: String? = nil,
                  textHash: String? = nil) -> Int? {
        return queue.sync {
            guard let db = db else { return nil }

            // 去重：检查是否与上一条文字记录相同
            if let hash = textHash {
                let checkSQL = """
                    SELECT text_hash FROM clipboard_items
                    WHERE content_type = 'text' ORDER BY id DESC LIMIT 1
                """
                var stmt: OpaquePointer?
                if sqlite3_prepare_v2(db, checkSQL, -1, &stmt, nil) == SQLITE_OK {
                    if sqlite3_step(stmt) == SQLITE_ROW,
                       let lastHash = sqlite3_column_text(stmt, 0) {
                        if String(cString: lastHash) == hash {
                            sqlite3_finalize(stmt)
                            return nil // 重复内容
                        }
                    }
                    sqlite3_finalize(stmt)
                }
            }

            let insertSQL = """
                INSERT INTO clipboard_items (content_type, text_content, image_path, text_hash)
                VALUES (?, ?, ?, ?)
            """
            var stmt: OpaquePointer?
            guard sqlite3_prepare_v2(db, insertSQL, -1, &stmt, nil) == SQLITE_OK else {
                print("[DatabaseService] prepare INSERT 失败")
                return nil
            }
            defer { sqlite3_finalize(stmt) }

            let typeStr = contentType.rawValue
            sqlite3_bind_text(stmt, 1, (typeStr as NSString).utf8String, -1, nil)

            if let text = textContent {
                sqlite3_bind_text(stmt, 2, (text as NSString).utf8String, -1, nil)
            } else {
                sqlite3_bind_null(stmt, 2)
            }

            if let path = imagePath {
                sqlite3_bind_text(stmt, 3, (path as NSString).utf8String, -1, nil)
            } else {
                sqlite3_bind_null(stmt, 3)
            }

            if let hash = textHash {
                sqlite3_bind_text(stmt, 4, (hash as NSString).utf8String, -1, nil)
            } else {
                sqlite3_bind_null(stmt, 4)
            }

            guard sqlite3_step(stmt) == SQLITE_DONE else {
                print("[DatabaseService] INSERT 失败")
                return nil
            }

            let newId = Int(sqlite3_last_insert_rowid(db))
            return newId
        }
    }

    // MARK: — 查询记录

    /// 获取剪贴板记录列表。置顶在前，按时间降序排列。
    func getItems(searchQuery: String = "", limit: Int = 200) -> [ClipboardItem] {
        return queue.sync {
            guard let db = db else { return [] }

            let sql: String
            if searchQuery.isEmpty {
                sql = """
                    SELECT * FROM clipboard_items
                    ORDER BY is_pinned DESC, created_at DESC
                    LIMIT ?
                """
            } else {
                sql = """
                    SELECT * FROM clipboard_items
                    WHERE (text_content LIKE ? OR content_type = 'image')
                    ORDER BY is_pinned DESC, created_at DESC
                    LIMIT ?
                """
            }

            var stmt: OpaquePointer?
            guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
                return []
            }
            defer { sqlite3_finalize(stmt) }

            if !searchQuery.isEmpty {
                let pattern = "%\(searchQuery)%"
                sqlite3_bind_text(stmt, 1, (pattern as NSString).utf8String, -1, nil)
                sqlite3_bind_int(stmt, 2, Int32(limit))
            } else {
                sqlite3_bind_int(stmt, 1, Int32(limit))
            }

            var items: [ClipboardItem] = []
            let dateFormatter = ISO8601DateFormatter()

            while sqlite3_step(stmt) == SQLITE_ROW {
                guard let item = parseItem(stmt: stmt!, dateFormatter: dateFormatter) else {
                    continue
                }
                items.append(item)
            }
            return items
        }
    }

    /// 根据 ID 获取单条记录
    func getItem(id: Int) -> ClipboardItem? {
        return queue.sync {
            guard let db = db else { return nil }

            let sql = "SELECT * FROM clipboard_items WHERE id = ?"
            var stmt: OpaquePointer?
            guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
                return nil
            }
            defer { sqlite3_finalize(stmt) }

            sqlite3_bind_int(stmt, 1, Int32(id))

            guard sqlite3_step(stmt) == SQLITE_ROW else { return nil }

            let dateFormatter = ISO8601DateFormatter()
            return parseItem(stmt: stmt!, dateFormatter: dateFormatter)
        }
    }

    /// 解析一行数据库记录为 ClipboardItem
    private func parseItem(stmt: OpaquePointer,
                           dateFormatter: ISO8601DateFormatter) -> ClipboardItem? {
        let id = Int(sqlite3_column_int(stmt, 0))

        guard let typeCStr = sqlite3_column_text(stmt, 1) else { return nil }
        let typeStr = String(cString: typeCStr)
        let contentType = ClipboardItem.ContentType(rawValue: typeStr) ?? .text

        var textContent: String? = nil
        if let textCStr = sqlite3_column_text(stmt, 2) {
            textContent = String(cString: textCStr)
        }

        var imagePath: String? = nil
        if let pathCStr = sqlite3_column_text(stmt, 3) {
            imagePath = String(cString: pathCStr)
        }

        // text_hash 列 (索引 4)，略过

        var createdAt = Date()
        if let dateCStr = sqlite3_column_text(stmt, 5) {
            // SQLite 存储格式：2026-07-20T12:34:56.789012
            var dateStr = String(cString: dateCStr)
            // 确保 ISO8601DateFormatter 可以解析
            if !dateStr.contains("T") {
                dateStr = dateStr.replacingOccurrences(of: " ", with: "T")
            }

            // ISO8601DateFormatter 需要精确格式
            if let date = dateFormatter.date(from: dateStr) {
                createdAt = date
            } else {
                // 尝试处理带微秒的格式
                let withZ = dateStr + "Z"
                if let date = dateFormatter.date(from: withZ) {
                    createdAt = date
                } else {
                    // 最后回退：手动解析
                    let trimmed = dateStr.components(separatedBy: ".").first ?? dateStr
                    if let date = dateFormatter.date(from: trimmed) {
                        createdAt = date
                    }
                }
            }
        }

        let isPinned = sqlite3_column_int(stmt, 6) != 0

        return ClipboardItem(
            id: id,
            contentType: contentType,
            textContent: textContent,
            imagePath: imagePath,
            createdAt: createdAt,
            isPinned: isPinned
        )
    }

    // MARK: — 置顶切换

    /// 切换置顶状态，返回新状态
    func togglePin(itemId: Int) -> Bool {
        return queue.sync {
            guard let db = db else { return false }

            // 查询当前状态
            let querySQL = "SELECT is_pinned FROM clipboard_items WHERE id = ?"
            var stmt: OpaquePointer?
            guard sqlite3_prepare_v2(db, querySQL, -1, &stmt, nil) == SQLITE_OK else {
                return false
            }
            defer { sqlite3_finalize(stmt) }

            sqlite3_bind_int(stmt, 1, Int32(itemId))
            guard sqlite3_step(stmt) == SQLITE_ROW else { return false }

            let current = sqlite3_column_int(stmt, 0)
            let newState: Int32 = current == 0 ? 1 : 0

            // 更新状态
            let updateSQL = "UPDATE clipboard_items SET is_pinned = ? WHERE id = ?"
            var updateStmt: OpaquePointer?
            guard sqlite3_prepare_v2(db, updateSQL, -1, &updateStmt, nil) == SQLITE_OK else {
                return false
            }
            defer { sqlite3_finalize(updateStmt) }

            sqlite3_bind_int(updateStmt, 1, newState)
            sqlite3_bind_int(updateStmt, 2, Int32(itemId))
            sqlite3_step(updateStmt)

            return newState != 0
        }
    }

    // MARK: — 删除记录

    /// 删除一条记录，同时清理磁盘图片文件
    func deleteItem(itemId: Int) {
        queue.sync {
            guard let db = db else { return }

            // 先查图片路径
            let querySQL = "SELECT image_path FROM clipboard_items WHERE id = ?"
            var stmt: OpaquePointer?
            if sqlite3_prepare_v2(db, querySQL, -1, &stmt, nil) == SQLITE_OK {
                sqlite3_bind_int(stmt, 1, Int32(itemId))
                if sqlite3_step(stmt) == SQLITE_ROW,
                   let pathCStr = sqlite3_column_text(stmt, 0) {
                    let imagePath = String(cString: pathCStr)
                    // 删除磁盘文件
                    if FileManager.default.fileExists(atPath: imagePath) {
                        try? FileManager.default.removeItem(atPath: imagePath)
                    }
                }
                sqlite3_finalize(stmt)
            }

            // 删除数据库记录
            let deleteSQL = "DELETE FROM clipboard_items WHERE id = ?"
            if sqlite3_prepare_v2(db, deleteSQL, -1, &stmt, nil) == SQLITE_OK {
                sqlite3_bind_int(stmt, 1, Int32(itemId))
                sqlite3_step(stmt)
                sqlite3_finalize(stmt)
            }
        }
    }

    // MARK: — 统计与清理

    /// 获取总记录数
    func getTotalCount() -> Int {
        return queue.sync {
            guard let db = db else { return 0 }
            var stmt: OpaquePointer?
            let sql = "SELECT COUNT(*) FROM clipboard_items"
            guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
                return 0
            }
            defer { sqlite3_finalize(stmt) }

            guard sqlite3_step(stmt) == SQLITE_ROW else { return 0 }
            return Int(sqlite3_column_int(stmt, 0))
        }
    }

    /// 清理过期的非置顶记录。返回清理条数。
    @discardableResult
    func cleanup(retentionDays: Int) -> Int {
        return queue.sync {
            guard let db = db else { return 0 }

            let cutoff = Date().addingTimeInterval(
                -Double(retentionDays) * 86400)
            let dateFormatter = ISO8601DateFormatter()
            let cutoffStr = dateFormatter.string(from: cutoff)

            // 先查要删除的图片路径
            let querySQL = """
                SELECT id, image_path FROM clipboard_items
                WHERE is_pinned = 0 AND created_at < ?
            """
            var stmt: OpaquePointer?
            guard sqlite3_prepare_v2(db, querySQL, -1, &stmt, nil) == SQLITE_OK else {
                return 0
            }

            sqlite3_bind_text(stmt, 1, (cutoffStr as NSString).utf8String, -1, nil)

            // 收集要删除的图片路径
            var imagePaths: [String] = []
            while sqlite3_step(stmt) == SQLITE_ROW {
                if let pathCStr = sqlite3_column_text(stmt, 1) {
                    imagePaths.append(String(cString: pathCStr))
                }
            }
            sqlite3_finalize(stmt)

            // 删除磁盘文件
            for path in imagePaths {
                if FileManager.default.fileExists(atPath: path) {
                    try? FileManager.default.removeItem(atPath: path)
                }
            }

            // 删除数据库记录
            let deleteSQL = """
                DELETE FROM clipboard_items
                WHERE is_pinned = 0 AND created_at < ?
            """
            guard sqlite3_prepare_v2(db, deleteSQL, -1, &stmt, nil) == SQLITE_OK else {
                return 0
            }
            defer { sqlite3_finalize(stmt) }

            sqlite3_bind_text(stmt, 1, (cutoffStr as NSString).utf8String, -1, nil)
            guard sqlite3_step(stmt) == SQLITE_DONE else { return 0 }

            let deleted = Int(sqlite3_changes(db))
            if deleted > 0 {
                print("[DatabaseService] 清理了 \(deleted) 条过期记录")
            }
            return deleted
        }
    }
}

File: windows/Services/DatabaseService.cs
using System.IO;
using Microsoft.Data.Sqlite;
using ClipboardManager.Models;

namespace ClipboardManager.Services;

/// <summary>
/// SQLite 数据库服务。
/// 与 Mac 版 DatabaseService 功能完全一致：
/// - 相同 schema、相同索引、相同 WAL 模式
/// - 线程安全（lock 串行化）
/// - 增删改查 + 清理过期记录
/// </summary>
public sealed class DatabaseService
{
    public static DatabaseService Instance { get; } = new();

    private readonly string _dbPath;
    private readonly string _imagesDir;
    private readonly Lock _lock = new(); // .NET 9 新 Lock 类型

    // ── 初始化 ──────────────────────────────────

    private DatabaseService() : this(GetDefaultDataDir()) { }

    /// <summary>测试用构造函数：指定数据目录</summary>
    internal DatabaseService(string dataDir)
    {
        Directory.CreateDirectory(dataDir);
        _dbPath = Path.Combine(dataDir, "history.db");
        _imagesDir = Path.Combine(dataDir, "images");
        Directory.CreateDirectory(_imagesDir);

        OpenAndInitialize();
    }

    private static string GetDefaultDataDir() =>
        Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "ClipboardManager");

    private void OpenAndInitialize()
    {
        using var db = new SqliteConnection($"Data Source={_dbPath}");
        db.Open();

        // WAL 模式提升并发性能
        using var pragmaCmd = db.CreateCommand();
        pragmaCmd.CommandText = "PRAGMA journal_mode=WAL";
        pragmaCmd.ExecuteNonQuery();

        // 建表（与 Mac 版完全一致）
        using var createCmd = db.CreateCommand();
        createCmd.CommandText = """
            CREATE TABLE IF NOT EXISTS clipboard_items (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                content_type TEXT NOT NULL,
                text_content TEXT,
                image_path TEXT,
                text_hash TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                is_pinned INTEGER DEFAULT 0
            );
            CREATE INDEX IF NOT EXISTS idx_created_at
                ON clipboard_items(created_at DESC);
            CREATE INDEX IF NOT EXISTS idx_pinned
                ON clipboard_items(is_pinned);
            """;
        createCmd.ExecuteNonQuery();

        System.Diagnostics.Debug.WriteLine(
            $"[DatabaseService] 数据库已打开: {_dbPath}");
    }

    // ── 公开属性 ─────────────────────────────────

    public string ImagesDir => _imagesDir;

    // ── 保存 ─────────────────────────────────────

    /// <summary>
    /// 保存一条记录。文字去重：比较上一条文字的 hash。
    /// 返回新记录 ID，重复则返回 null。
    /// </summary>
    public int? SaveItem(ContentType contentType,
                         string? textContent = null,
                         string? imagePath = null,
                         string? textHash = null)
    {
        lock (_lock)
        {
            using var db = new SqliteConnection($"Data Source={_dbPath}");
            db.Open();

            // 去重：检查文字 hash 是否与上一条相同
            if (contentType == ContentType.Text && textHash != null)
            {
                using var checkCmd = db.CreateCommand();
                checkCmd.CommandText = """
                    SELECT text_hash FROM clipboard_items
                    WHERE content_type = 'text'
                    ORDER BY id DESC LIMIT 1
                    """;
                var lastHash = checkCmd.ExecuteScalar() as string;
                if (lastHash == textHash)
                    return null; // 重复，跳过
            }

            // 插入
            using var insertCmd = db.CreateCommand();
            insertCmd.CommandText = """
                INSERT INTO clipboard_items
                    (content_type, text_content, image_path, text_hash)
                VALUES
                    (@type, @text, @image, @hash)
                """;

            insertCmd.Parameters.AddWithValue("@type",
                contentType == ContentType.Text ? "text" : "image");
            insertCmd.Parameters.AddWithValue("@text",
                (object?)textContent ?? DBNull.Value);
            insertCmd.Parameters.AddWithValue("@image",
                (object?)imagePath ?? DBNull.Value);
            insertCmd.Parameters.AddWithValue("@hash",
                (object?)textHash ?? DBNull.Value);

            insertCmd.ExecuteNonQuery();

            // 获取新插入 ID
            using var lastIdCmd = db.CreateCommand();
            lastIdCmd.CommandText = "SELECT last_insert_rowid()";
            var newId = Convert.ToInt32(lastIdCmd.ExecuteScalar()!);
            return newId;
        }
    }

    // ── 查询 ─────────────────────────────────────

    /// <summary>
    /// 获取记录列表。置顶在前，时间降序。
    /// 搜索时匹配 text_content LIKE（图片始终保留）。
    /// </summary>
    public List<ClipboardItem> GetItems(string searchQuery = "", int limit = 200)
    {
        lock (_lock)
        {
            using var db = new SqliteConnection($"Data Source={_dbPath}");
            db.Open();

            using var cmd = db.CreateCommand();
            if (string.IsNullOrWhiteSpace(searchQuery))
            {
                cmd.CommandText = """
                    SELECT * FROM clipboard_items
                    ORDER BY is_pinned DESC, created_at DESC
                    LIMIT @limit
                    """;
                cmd.Parameters.AddWithValue("@limit", limit);
            }
            else
            {
                cmd.CommandText = """
                    SELECT * FROM clipboard_items
                    WHERE (text_content LIKE @pattern OR content_type = 'image')
                    ORDER BY is_pinned DESC, created_at DESC
                    LIMIT @limit
                    """;
                cmd.Parameters.AddWithValue("@pattern", $"%{searchQuery}%");
                cmd.Parameters.AddWithValue("@limit", limit);
            }

            var items = new List<ClipboardItem>();
            using var reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                items.Add(ParseItem(reader));
            }
            return items;
        }
    }

    /// <summary>根据 ID 获取单条记录（复制时使用）</summary>
    public ClipboardItem? GetItem(int id)
    {
        lock (_lock)
        {
            using var db = new SqliteConnection($"Data Source={_dbPath}");
            db.Open();

            using var cmd = db.CreateCommand();
            cmd.CommandText = "SELECT * FROM clipboard_items WHERE id = @id";
            cmd.Parameters.AddWithValue("@id", id);

            using var reader = cmd.ExecuteReader();
            return reader.Read() ? ParseItem(reader) : null;
        }
    }

    /// <summary>解析行数据为 ClipboardItem</summary>
    private static ClipboardItem ParseItem(SqliteDataReader reader)
    {
        // 列顺序：id(0), content_type(1), text_content(2), image_path(3),
        //          text_hash(4), created_at(5), is_pinned(6)
        return new ClipboardItem
        {
            Id = reader.GetInt32(0),
            Type = reader.GetString(1) == "text"
                ? ContentType.Text : ContentType.Image,
            TextContent = reader.IsDBNull(2) ? null : reader.GetString(2),
            ImagePath = reader.IsDBNull(3) ? null : reader.GetString(3),
            CreatedAt = ParseDateTime(reader.GetString(5)),
            IsPinned = reader.GetInt32(6) != 0
        };
    }

    /// <summary>
    /// 解析 SQLite 日期。支持多种格式：
    /// - ISO8601 with T: 2026-07-20T12:34:56
    /// - ISO8601 with microseconds: 2026-07-20T12:34:56.789012
    /// - Space format: 2026-07-20 12:34:56
    /// </summary>
    private static DateTime ParseDateTime(string dateStr)
    {
        // 统一处理
        var normalized = dateStr.Replace(' ', 'T');

        // 去掉微秒（如果存在）
        var dotIndex = normalized.IndexOf('.');
        if (dotIndex > 0)
        {
            normalized = normalized[..dotIndex];
        }

        if (DateTime.TryParse(normalized, out var result))
            return result;

        return DateTime.Now; // 最终回退
    }

    // ── 置顶切换 ─────────────────────────────────

    /// <summary>切换置顶状态，返回新状态</summary>
    public bool TogglePin(int itemId)
    {
        lock (_lock)
        {
            using var db = new SqliteConnection($"Data Source={_dbPath}");
            db.Open();

            // 查询当前状态
            using var queryCmd = db.CreateCommand();
            queryCmd.CommandText =
                "SELECT is_pinned FROM clipboard_items WHERE id = @id";
            queryCmd.Parameters.AddWithValue("@id", itemId);
            var current = Convert.ToInt32(queryCmd.ExecuteScalar()!);

            // 翻转
            var newState = current == 0 ? 1 : 0;
            using var updateCmd = db.CreateCommand();
            updateCmd.CommandText =
                "UPDATE clipboard_items SET is_pinned = @state WHERE id = @id";
            updateCmd.Parameters.AddWithValue("@state", newState);
            updateCmd.Parameters.AddWithValue("@id", itemId);
            updateCmd.ExecuteNonQuery();

            return newState != 0;
        }
    }

    // ── 删除 ─────────────────────────────────────

    /// <summary>
    /// 删除一条记录。如果是图片，同时删除磁盘文件。
    /// 与 Mac 版行为一致。
    /// </summary>
    public void DeleteItem(int itemId)
    {
        lock (_lock)
        {
            using var db = new SqliteConnection($"Data Source={_dbPath}");
            db.Open();

            // 查询图片路径（如有）
            using var queryCmd = db.CreateCommand();
            queryCmd.CommandText =
                "SELECT image_path FROM clipboard_items WHERE id = @id";
            queryCmd.Parameters.AddWithValue("@id", itemId);
            var imagePath = queryCmd.ExecuteScalar() as string;

            // 删除磁盘文件
            if (!string.IsNullOrEmpty(imagePath) && File.Exists(imagePath))
            {
                try { File.Delete(imagePath); }
                catch (Exception ex)
                {
                    System.Diagnostics.Debug.WriteLine(
                        $"[DatabaseService] 删除图片失败: {ex.Message}");
                }
            }

            // 删除数据库记录
            using var deleteCmd = db.CreateCommand();
            deleteCmd.CommandText =
                "DELETE FROM clipboard_items WHERE id = @id";
            deleteCmd.Parameters.AddWithValue("@id", itemId);
            deleteCmd.ExecuteNonQuery();
        }
    }

    // ── 统计 ─────────────────────────────────────

    /// <summary>获取总记录数（用于状态栏显示）</summary>
    public int GetTotalCount()
    {
        lock (_lock)
        {
            using var db = new SqliteConnection($"Data Source={_dbPath}");
            db.Open();

            using var cmd = db.CreateCommand();
            cmd.CommandText = "SELECT COUNT(*) FROM clipboard_items";
            return Convert.ToInt32(cmd.ExecuteScalar()!);
        }
    }

    // ── 清理 ─────────────────────────────────────

    /// <summary>
    /// 清理过期的非置顶记录。返回清理条数。
    /// 与 Mac 版行为一致：先删图片文件，再删数据库记录。
    /// </summary>
    public int Cleanup(int retentionDays)
    {
        lock (_lock)
        {
            using var db = new SqliteConnection($"Data Source={_dbPath}");
            db.Open();

            var cutoff = DateTime.Now.AddDays(-retentionDays);
            var cutoffStr = cutoff.ToString("yyyy-MM-ddTHH:mm:ss");

            // 收集要删除的图片路径
            var imagePaths = new List<string>();
            using (var queryCmd = db.CreateCommand())
            {
                queryCmd.CommandText = """
                    SELECT image_path FROM clipboard_items
                    WHERE is_pinned = 0 AND created_at < @cutoff
                      AND image_path IS NOT NULL
                    """;
                queryCmd.Parameters.AddWithValue("@cutoff", cutoffStr);

                using var reader = queryCmd.ExecuteReader();
                while (reader.Read())
                {
                    if (!reader.IsDBNull(0))
                        imagePaths.Add(reader.GetString(0));
                }
            }

            // 删除磁盘文件
            foreach (var path in imagePaths)
            {
                if (File.Exists(path))
                {
                    try { File.Delete(path); }
                    catch (Exception ex)
                    {
                        System.Diagnostics.Debug.WriteLine(
                            $"[DatabaseService] 清理图片失败: {ex.Message}");
                    }
                }
            }

            // 删除数据库记录
            using var deleteCmd = db.CreateCommand();
            deleteCmd.CommandText = """
                DELETE FROM clipboard_items
                WHERE is_pinned = 0 AND created_at < @cutoff
                """;
            deleteCmd.Parameters.AddWithValue("@cutoff", cutoffStr);
            var deleted = deleteCmd.ExecuteNonQuery();

            if (deleted > 0)
                System.Diagnostics.Debug.WriteLine(
                    $"[DatabaseService] 清理了 {deleted} 条过期记录");

            return deleted;
        }
    }
}
